onnxruntime/core/optimizer/gemm_sum_fusion.cc

```diff
@@ -0,0 +1,161 @@
+  // Also (maybe) get the sum node's other input edge, unless it's a graph input.
+  if (sum_node.InputDefs()[0]->Name() == gemm_node.OutputDefs()[0]->Name()) {
+    new_gemm_input_defs.push_back(sum_node.MutableInputDefs()[1]);
+  } else {
+    new_gemm_input_defs.push_back(sum_node.MutableInputDefs()[0]);
+  }
+  ORT_ENFORCE(new_gemm_input_defs.size() == 3);
+
+  std::vector<NodeArg*> new_gemm_output_defs = sum_node.MutableOutputDefs();
+  ORT_ENFORCE(new_gemm_output_defs.size() == 1);
+
+  Node& new_gemm_node = graph.AddNode(graph.GenerateNodeName(gemm_node.Name() + "_sum_transformed"),
+                                      gemm_node.OpType(),
+                                      "Fused Gemm with Sum",
+                                      new_gemm_input_defs,
+                                      new_gemm_output_defs,
+                                      {},
+                                      gemm_node.Domain());
+  new_gemm_node.AddAttribute("transA", static_cast<int64_t>(transA));
+  new_gemm_node.AddAttribute("transB", static_cast<int64_t>(transB));
+  new_gemm_node.AddAttribute("alpha", static_cast<float>(alpha));
+  new_gemm_node.AddAttribute("beta", static_cast<float>(beta));
+
+  // Move both input edges from original gemm to new gemm.
+  for (auto input_edge : graph_utils::GraphEdge::GetNodeInputEdges(gemm_node)) {
+    ORT_ENFORCE(input_edge.src_arg_index < 2);
+    graph.AddEdge(input_edge.src_node, new_gemm_node.Index(), input_edge.src_arg_index, input_edge.dst_arg_index);
+    graph.RemoveEdge(input_edge.src_node, input_edge.dst_node, input_edge.src_arg_index, input_edge.dst_arg_index);
+  }
+
+  // Move all output edges from sum to new gemm.
+  for (auto output_edge : graph_utils::GraphEdge::GetNodeOutputEdges(sum_node)) {
+    ORT_ENFORCE(output_edge.src_arg_index == 0);
+    graph.AddEdge(new_gemm_node.Index(), output_edge.dst_node, output_edge.src_arg_index, output_edge.dst_arg_index);
+    graph.RemoveEdge(output_edge.src_node, output_edge.dst_node, output_edge.src_arg_index, output_edge.dst_arg_index);
+  }
+
+  // Finally, move the other sum input edge to "C" for the new gemm node.
+  // If The other sum input def is a a graph input, there is no edge to move.
+  bool sum_input_moved = false;
+  for (auto sum_input : graph_utils::GraphEdge::GetNodeInputEdges(sum_node)) {
+    if (sum_input.src_node != gemm_node.Index()) {
+      ORT_ENFORCE(!sum_input_moved);
+      graph.AddEdge(sum_input.src_node, new_gemm_node.Index(), sum_input.src_arg_index, 2);
+      graph.RemoveEdge(sum_input.src_node, sum_input.dst_node, sum_input.src_arg_index, sum_input.dst_arg_index);
+      sum_input_moved = true;
+    }
+  }
+  // Old gemm node output is no longer needed. It was previously fed into the
+  // sum node which is now also handled by the new gemm. Remove this output edge
+  // to allow the node to be removed from the graph.
+  graph_utils::RemoveNodeOutputEdges(graph, gemm_node);
```

Comment:
```suggestion
  graph_utils::RemoveNodeOutputEdges(graph, gemm_node);
  graph_utils::RemoveNodeOutputEdges(graph, sum_node);
```

Reply:
Sum node outputs should have already been moved (and if there's any left, it would indicate a logic error). I was thinking that simply removing all output edges could possibly mask a logic error, if one were to exist. Thoughts on the following instead?

```c++
  // All output edges of the sum node should have been removed already, so we can
  // remove it from the graph.
  ORT_ENFORCE(sum_node.GetOutputEdgesCount() == 0);
  ORT_ENFORCE(graph.RemoveNode(sum_node.Index()));
```

Reply:
Looks good to me. Thanks for the explanation.